addons/event_crm/models/event_registration.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+
+
+class EventRegistration(models.Model):
+    _inherit = 'event.registration'
+
+    lead_ids = fields.Many2many('crm.lead', string='Leads', readonly=True, copy=False,
+        help="Leads generated from the registration.")
+    lead_count = fields.Integer('# Leads', compute='_compute_lead_count',
+        help="Counter for the leads linked to this registration")
+
+    @api.depends('lead_ids')
+    def _compute_lead_count(self):
+        for record in self:
+            record.lead_count = len(record.lead_ids)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        registrations = super(EventRegistration, self).create(vals_list)
+        self.env['event.lead.rule'].search([]).sudo()._create_leads(registrations)
+        return registrations
+
+    def write(self, vals):
+        registration = super(EventRegistration, self).write(vals)
+        if vals.get('partner_id') or vals.get('main_registration_id') or vals.get('event_id'):
+            for lead in self.lead_ids.sorted('id'):
+                vals = lead.registration_ids._get_registration_lead_values(lead.registration_rule_id)
+                lead.write(vals)
+        return registration
+
+    def _get_registration_lead_values(self, rule):
+        registration_lead_values = {
+            'user_id': rule.lead_user_id.id,
+            'type': rule.lead_type,
+            'team_id': rule.lead_sales_team_id.id,
+            'tag_ids': rule.lead_tag_ids.ids,
+            'registration_rule_id': rule.id,
+            'event_id': self.event_id.id,
+            'referred': self.event_id.name,
+            'registration_ids': self.ids,
+        }
+        if rule.lead_creation_basis == 'attendee':
+            description = _("Participant:\n")
+            description += self._get_lead_description()
+            registration_lead_values.update({
+                'description': description,
+            })
+
+            if self.partner_id and self.partner_id != self.env.ref('base.public_partner'):
+                leads = self.main_registration_id.lead_ids.filtered(lambda lead: lead.registration_rule_id == rule)
```

Comment:
Why do you update the main registration when being per attendee rule ? It seems you make some spaghetti with _create_leads and _get_registration_lead_values that are not clearly separated and values are updated a bit everytime, everywhere.